spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowServerConfiguration.java

```diff
@@ -161,10 +212,77 @@ public void configurePathMatch(PathMatchConfigurer configurer) {
 	}
 
 	@Bean
-	@ConditionalOnMissingBean
-	public TaskExplorer taskExplorer(DataSource dataSource) {
-		JdbcTaskExplorerFactoryBean factoryBean = new JdbcTaskExplorerFactoryBean(dataSource);
-		return factoryBean.getObject();
+	public TaskExplorerFactoryBean taskExplorerFactoryBean(DataSource dataSource) {
+		return new TaskExplorerFactoryBean(dataSource);
+	}
+
+	@Bean
+	public TaskJobRepository taskJobExecutionRepository(JobService service, TaskExplorer taskExplorer) {
+		return new TaskJobRepository(service, taskExplorer);
+	}
+
+	@Bean
+	public SimpleJobServiceFactoryBean simpleJobServiceFactoryBean(DataSource dataSource,
+			JobRepositoryFactoryBean repositoryFactoryBean) throws Exception {
+		SimpleJobServiceFactoryBean factoryBean = new SimpleJobServiceFactoryBean();
+		factoryBean.setDataSource(dataSource);
+		factoryBean.setJobRepository(repositoryFactoryBean.getObject());
+		factoryBean.setJobLocator(new MapJobRegistry());
+		factoryBean.setJobLauncher(new SimpleJobLauncher());
+		factoryBean.setDataSource(dataSource);
+		return factoryBean;
+	}
+
+	@Bean
+	public JobExplorerFactoryBean jobExplorerFactoryBean(DataSource dataSource) {
+		JobExplorerFactoryBean jobExplorerFactoryBean = new JobExplorerFactoryBean();
+		jobExplorerFactoryBean.setDataSource(dataSource);
+		return jobExplorerFactoryBean;
+	}
+
+	@Bean
+	@ConditionalOnExpression("#{'${spring.datasource.url:}'.startsWith('jdbc:h2:tcp://localhost:') && '${spring.datasource.url:}'.contains('/mem:')}")
```

Comment:
For all of these that are conditional on expression, instead of duplicating that expression, why not create an inner class and add it to that?